benchmarks/gpt_fast/benchmark.py

```diff
@@ -18,57 +21,172 @@ class Experiment:
+            )
+
+        memory_bandwidth = memory_bandwidth / len(input_shapes)
+        dtype_str = str(dtype).replace("torch.", "")
+        results.append(
+            Experiment(
+                f"mlp_layer_norm_gelu_{dtype_str}",
+                "memory_bandwidth(GB/s)",
+                expected_memory_bandwidth,
+                f"{memory_bandwidth:.02f}",
+            )
+        )
+    return results
+
+
+def run_layer_norm():
+    dtype_memory_bandwidth_map = {
+        torch.bfloat16: "1050",
+    }
+    input_shapes = [1024, 4096, 8192, 16384]
+    BS = 4096
+    results = []
+    for dtype, expected_memory_bandwidth in dtype_memory_bandwidth_map.items():
+        memory_bandwidth = 0
+        for D in input_shapes:
+            mod = nn.LayerNorm(D).to("cuda")
+
+            x = torch.randn(BS, D, device="cuda", dtype=dtype)
+
+            compiled_mod = torch.compile(mod)
+
+            for _ in range(WARMUP_ITER):
+                compiled_mod(x)
 
-    for i in range(start, num_samples):
-        torch.cuda.synchronize("cuda")
+            us_per_iter = do_bench(lambda: compiled_mod(x)) * 1000
+            memory_bandwidth += (1e6 / us_per_iter) * 2 * BS * D * dtype.itemsize / 1e9
 
-        t0 = time.perf_counter()
-        mod(x)
+        memory_bandwidth = memory_bandwidth / len(input_shapes)
+        dtype_str = str(dtype).replace("torch.", "")
+        results.append(
+            Experiment(
+                f"layer_norm_{dtype_str}",
+                "memory_bandwidth(GB/s)",
+                expected_memory_bandwidth,
+                f"{memory_bandwidth:.02f}",
+            )
+        )
+    return results
+
+
+def run_gather_gemv():
+    E = 8
+    dtype_memory_bandwidth_map = {
+        torch.int8: "1195",
+        torch.bfloat16: "2180",
+    }
+    input_shapes = [1024, 4096, 8192, 16384]
+    results = []
+    for dtype, expected_memory_bandwidth in dtype_memory_bandwidth_map.items():
+        memory_bandwidth = 0
+        for D in input_shapes:
 
-        if i == -1:
-            print(f"Compilation time: {time.perf_counter() - t0:.2f} seconds")
-            continue
+            def gather_gemv(W, score_idxs, x):
+                return W[score_idxs].to(x.dtype) @ x
 
-        torch.cuda.synchronize("cuda")
-        total_time += time.perf_counter() - t0
+            W = torch.randn(E, D, D, device="cuda").to(dtype=dtype)
+            x = torch.randn(D, device="cuda", dtype=torch.bfloat16)
+            score_idxs = torch.tensor([3, 5], device="cuda")
 
-    total_time = total_time / num_samples
+            compiled_fn = torch.compile(gather_gemv)
 
-    return total_time
+            for _ in range(WARMUP_ITER):
+                compiled_fn(W, score_idxs, x)
 
+            us_per_iter = do_bench(lambda: compiled_fn(W, score_idxs, x)) * 1000
+            memory_bandwidth += (1e6 / us_per_iter) * 4 * D * D * dtype.itemsize / 1e9
```

Comment:
Why is this `4 * D* D`? I think it should be 2 * D * D.

Reply:
both read and write?

Reply:
Addressed offline.